frontend/app/services/ha_pair/ha_pair.js

```diff
@@ -0,0 +1,10 @@
+(function () {
+    angular.module('workshop')
+        .service('HAPairService', HAPairService);
+    HAPairService.$inject = ['$http'];
+    function HAPairService($http) {
+        this.add_ha_pair = function (ha_pair) {
+            return $http.post('/api/hapair', ha_pair);
```

Comment:
Nice! The takeaway from here is that the $http service returns a promise, you return it from this.add_ha_pair function and then you resolve it in your controller.